File: gittools/models/commit.py

from scriptcore.basescript import BaseScript


class Commit(object):

    @staticmethod
    def all(limit=10):
        """
        Get all commits
        :param limit:   The limit
        :return:        List of Commit's
        """

        current_script = BaseScript.current_script()
        output = current_script.execute('git log -n %i --pretty=format:"%%H | \%%h | %%ai | %%an <%%ae> | %%s"' % limit)

        commits = []
        for line in output:
            parts = line.split(' | ')

            commits.append(Commit(parts[0], parts[1], parts[3], parts[4], parts[2]))

        return commits

    @staticmethod
    def where(where, limit=10):
        """
        Get filtered commits
        :param where:   Where statement
        :param limit:   The limit
        :return:        List of Commit's
        """

        current_script = BaseScript.current_script()
        output = current_script.execute('git log -n %i %s --pretty=format:"%%H | \%%h | %%ai | %%an <%%ae> | %%s"' % (limit, where))

        commits = []
        for line in output:
            parts = line.split(' | ')

            commits.append(Commit(parts[0], parts[1], parts[3], parts[4], parts[2]))

        return commits

    def __init__(self, full_hash, short_hash, author, message, datetime):
        """
        Constructor
        :param full_hash:   The full hash
        :param short_hash:  The short hash
        :param author:      The author
        :param message:     The message
        :param datetime:    The date time
        """
        self.full_hash = full_hash
        self.short_hash = short_hash
        self.author = author
        self.message = message
        self.datetime = datetime

File: src/models/commit.ts
import * as childProcess from 'child_process'
import * as osLocale from 'os-locale'

export default class Commit {
  fullHash: string
  shortHash: string
  author: string
  message: string
  datetime: string

  static all (limit: number): Commit[] {
    const output = childProcess.execSync(`git log -n ${limit} --pretty=format:"%H | %h | %ai | %an <%ae> | %s"`).toString()

    return output.split('\n')
      .map(line => {
        const parts = line.split(' | ')
        return new Commit(parts[0], parts[1], parts[3], parts[4], parts[2])
      })
  }

  static where (where: string, limit: number): Commit[] {
    const output = childProcess.execSync(`git log -n ${limit} ${where} --pretty=format:"%H | %h | %ai | %an <%ae> | %s"`).toString()

    return output.split('\n')
      .map(line => {
        const parts = line.split(' | ')
        return new Commit(parts[0], parts[1], parts[3], parts[4], parts[2])
      })
  }

  constructor (fullHash: string, shortHash: string, author: string, message: string, datetime: string) {
    this.fullHash = fullHash
    this.shortHash = shortHash
    this.author = author
    this.message = message
    this.datetime = datetime
  }

  get date (): Date {
    return new Date(this.datetime)
  }

  get formattedDate (): string {
    const locale = osLocale.sync().replace('_', '-')
    return this.date.toLocaleTimeString(locale)
  }
}
